### PR TITLE
Orphans and partial collections

### PR DESCRIPTION
**Pull request by alexdunnjpl**
### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [x] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [ ] AI generated substantial portions of this code

Estimated % of code influenced by AI: ___ %

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### Terraform (only if this PR touches a `terraform/` directory)

- [ ] **Validator run:** `scripts/validate_terraform.py` was run against the changed `terraform/` tree and any Must-Have failures were resolved (or are tracked in a reviewed `.tfvalidate-ignore` entry, not silently bypassed).
- [ ] **Ownership boundary:** Shared/singleton CDS resources (Cognito, CloudFront, org-wide IAM roles, shared security groups) were only added/changed in `pdc-cds-infra`, not duplicated in an application repo.
- [ ] **Module structure:** `modules/` stays flat and no new module is just a thin wrapper around a single resource.
- [ ] **SSM interface:** Any new cross-component output is actually published under `/pds/<component>/...` and, if this PR is the consumer side, reads from the correct published parameter rather than another repo's Terraform state.
- [ ] **Naming intent:** Resource and variable names are semantically meaningful, not just non-redundant (the validator only checks for repeated resource-type substrings).
- [ ] **Auth at runtime:** Terraform/CI actually authenticates via OIDC or an assumed role when applied — not just "no static key found in this diff."
- [ ] **Least privilege:** Any new or changed IAM policy attached to the Terraform execution role grants only what this change needs.

### CODE REVIEW

**Comment by alexdunnjpl** [comment]
@copilot resolve the CHANGELOG.md merge conflicts in this pull request

File: scratch/bad-doc-id-enumerator/enum.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                if search_after is not None:
+                    body["search_after"] = search_after
+
+                resp = retry_call(lambda: client.search(body=body), tries=9, delay=2, backoff=2)
+                hits = resp["hits"]["hits"]
+                total_read += len(hits)
+                print(f'STATUS: {total_read} docs checked')
+
+                if not hits:
+                    break
```
Comment (nutjob4life) [inline scratch/bad-doc-id-enumerator/enum.py:76]: This `break` statement prevents anything else from being processed, right? When `hits` is nonempty, none of the code executes either so `search_after` remains `None`.

I think the `for hit in hits…` and everything beneath it should be de-indented, right?

File: src/pds/registrysweepers/utils/db/__init__.py [hunk 2]
```diff
@@ -475,3 +541,31 @@ def get_query_hits_count(client: OpenSearch, index_name: str, query: Dict) -> in
...
+        _generate_delete_actions(),
+        chunk_size=chunk_size,
+        raise_on_error=False,
+        raise_on_exception=True,
+        max_retries=3,
+    )
+
+    non_404_errors = [e.get("delete", {}).get("status") != 404 for e in errors]
+    if non_404_errors:
+        raise RuntimeError(f'Some deletions have failed on index {index}: {", ".join(non_404_errors)}')
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/__init__.py]: Note that `non_404_errors` is not a list of errors, but a list of booleans, i.e., something like

    [False, True, False]

Later this `join`

    ", ".join(non_404_errors)

will raise

    TypeError: sequence item 0: expected str instance, bool found

so any delete errors, including the 404s we're trying to tolerate, blow up the error handler.

File: src/pds/registrysweepers/ancestry/updatedeferraltracker.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+       3. A later, different update for that document is processed in the same stream and is marked for release.
+
+       In this case, it is necessary to avoid releasing the earlier update as it has not been written successfully.
+
+    N.B. Not thread safe
+    """
+    _source: Iterable[Update]
+    _temp_store: dict[str, List[Update]] = defaultdict(list)
+    _store: dict[str, List[Update]] = defaultdict(list)
+
```
Comment (nutjob4life) [inline src/pds/registrysweepers/ancestry/updatedeferraltracker.py:31]: Note that `_source`, `_temp_store`, and `_store` are _class attributes_, so all tracker instances initially share `_temp_store` and `_store`.

The `__init__` c'tor adds an instance attribute for `_source`, but not the others.

This appears to work because some of the methods eventually assign instance attributes and `flush()` drains things; but exceptions, overlapping isntances, tests, or future reuse can leak deferred updates between trackers.

You probably want:
```python
def __init__(self, source: Iterable[Update]):
    self._source = source
    self._temp_store = defaultdict(list)
    self._store = defaultdict(list)
```
right?

File: src/pds/registrysweepers/ancestry/main.py [hunk 4]
```diff
@@ -75,40 +79,50 @@ def run(
...
+            pending_updates,
+            index_name=resolve_multitenant_index_name(client, "registry"),
+        )
+
+        # Force a refresh to prevent read-after-write consistency from hiding the applied updates
+        client.indices.refresh(index=registry_index_name)
+
+        remaining_orphaned_docs = get_orphaned_documents(client, registry_index_name)
+        remaining_orphaned_doc_ids = {doc.get("_id") for doc in remaining_orphaned_docs}
+        successfully_merged_doc_ids = pending_update_doc_ids.difference(remaining_orphaned_doc_ids)
```
Comment (nutjob4life) [inline src/pds/registrysweepers/ancestry/main.py]: It looks like we're using `pending_update_doc_ids` and subtracting the `remaining_orphaned_doc_ids` from it, not from the current `batch`. Two issues:

1. Previously successful IDs are rediscovered and counted on _every_ subsequent iteration:

    successfully_merged_doc_count += len(successfully_merged_doc_ids)

Suppose there 2500 documents. After batch 1 you could count 1000. After batch 2 you could count 2000 more. And after batch 3 you could count 2500 more.

The log announces 5500 successful merges for 2500 docs.

2. We repeatedly attempt to delete already-deleted deferred docs, producing exactly the 404 errors that `bulk_delete_documents()` was supposed to tolerate (aside from the fact that its own 404 handling may be broken).

Maybe this should be?
```python
batch_ids = set(batch)

successfully_merged_doc_ids = batch_ids.difference(
    remaining_orphaned_doc_ids
)
```

File: src/pds/registrysweepers/ancestry/queries.py [hunk 5]
```diff
@@ -149,3 +116,13 @@ def get_orphaned_documents_count(client: OpenSearch, index_name: str) -> int:
     # Query an index documents without an up-to-date ancestry version reference - this would indicate a product which is
     # orphaned and is getting missed in processing
     return get_query_hits_count(client, index_name, _orphaned_docs_query)
+
+
+def get_deferred_update_documents(client: OpenSearch, index_name: str) -> Iterable[Dict]:
+    deferred_updates_index_name = f"{index_name}-deferred-updates"
```
Comment (nutjob4life) [inline src/pds/registrysweepers/ancestry/queries.py:122]: This variable `deferred_updates_index_name` is unused.

File: README.md [hunk 6]
```diff
@@ -7,13 +7,17 @@ This package provides supplementary metadata generation for registry documents,
...
 
-[Accepts environment variables to tune performance](./src/pds/registrysweepers/ancestry/runtimeconstants.py), primarily trading increased runtime duration for reduced peak memory usage.
+The ancestry sweeper generates membership metadata for each product, i.e. which bundle and collection identifiers (LID
+and LIDVID) reference a given product. These values will be stored in property
+`[[ops:Provenance/ops:parent_bundle_identifier` and `ops:Registry_Sweepers.ops:ancestor_refs)`.
+ <TODO: describe mechanism after reimplementation is complete - edunn 20260630>
```
Comment (nutjob4life) [inline README.md]: `[[` might be accidental? It mentions `ops:Provenance/...` while the PR is migrating away from that. The opening backticks/brackets/parens don't balance.

But that raw internal `TODO` really sticks out in a user-facing README 😉

File: tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py [hunk 7]
```diff
@@ -110,6 +110,7 @@ def test_empty_results_completes_successfully(self, mock_opensearch_client):
         # Should not generate any updates
         assert len(bulk_updates) == 0
 
+    @pytest.mark.skip(reason='Requires reimplementation or removal as mock is incompatible with https://github.com/NASA-PDS/registry-sweepers/pull/256')
```
Comment (nutjob4life) [inline tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py]: Should we keep skipping this test? Ancestry gets major changes in this PR and it might be nicer to have tests for it.

File: src/pds/registrysweepers/utils/db/__init__.py [hunk 8]
```diff
@@ -290,17 +296,35 @@ def mock_wrapper(
...
-    as_upsert: bool = False,
+        client: OpenSearch,
+        updates: Iterable[Update],
+        index_name: str,
+        bulk_chunk_max_update_count: Union[int, None] = None,
+        as_upsert: bool = False,
+        defer_on_missing_document=False
 ):
+    if as_upsert and defer_on_missing_document:
+        raise RuntimeError('"as_upsert" and "defer_if_not_exists" may not both be set as they are mutually-exclusive')
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/__init__.py]: `defer_if_not_exists` should be `defer_on_missing_document`, right?

**Review by nutjob4life — CHANGES_REQUESTED** [review]
Whew! Wow, this is enormous. Thanks @alexdunnjpl for this incredible effort. I did notice a few things that we might want to clear up before proceeding. Take a look?

File: scratch/bad-doc-id-enumerator/enum.py (hunk 1 above)
Comment (alexdunnjpl) [reply scratch/bad-doc-id-enumerator/enum.py:76]: Right you are - pasted an update from vim and missed that, thanks!

File: src/pds/registrysweepers/ancestry/main.py (hunk 4 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/ancestry/main.py]: Good catches, thanks!

1. Looking with fresh eyes, this whole part need not be iterative.

2. Will fix

File: src/pds/registrysweepers/ancestry/queries.py (hunk 5 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/ancestry/queries.py:122]: oof, thanks Sean.  This resulted in the return of all documents rather than just what had updates, which passed testing but would have been very slow and problematic at scale.

File: src/pds/registrysweepers/ancestry/updatedeferraltracker.py (hunk 3 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/ancestry/updatedeferraltracker.py:31]: right!

File: tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py (hunk 7 above)
Comment (alexdunnjpl) [reply tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py]: In this case, yes - will hit you up in slack

File: src/pds/registrysweepers/utils/db/__init__.py (hunk 2 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/__init__.py]: will fix

File: src/pds/registrysweepers/utils/db/__init__.py (hunk 8 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/__init__.py]: yep, tyvm

File: tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py (hunk 7 above)
Comment (alexdunnjpl) [reply tests/pds/registrysweepers/ancestry/integration/test_full_pipeline.py]: Actually, this test should be removed completely, as it's already tested here https://github.com/NASA-PDS/registry-sweepers/blob/e2b0946ed6226ad8cb4553ac004b6107f2a1c2ce/tests/pds/registrysweepers/ancestry/unit/test_utils_conversion.py#L40

File: README.md (hunk 6 above)
Comment (alexdunnjpl) [reply README.md]: will fix

**Review by nutjob4life — APPROVED** [review]
Original Δs: ✓
New Δs: ✓
Tox: ✓
Approval: ✅
Cause for celebration: 🥳
Tox details:
```
  py312: OK (6.02=setup[1.68]+cmd[4.34] seconds)
  py313: OK (5.03=setup[0.84]+cmd[4.19] seconds)
  docs: OK (1.17=setup[0.57]+cmd[0.60] seconds)
  lint: OK (4.43=setup[0.01]+cmd[4.42] seconds)
  congratulations :) (16.85 seconds)
```

File: scratch/bad-doc-id-enumerator/enum.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
+"""
```
Comment (jordanpadams) [inline scratch/bad-doc-id-enumerator/enum.py:1]: @alexdunnjpl we can leave this for now, but I would like us to standardizing on naming of directories like this. in other repos, this is a `scripts` directory. I feel that it makes sense for us to rename this repo to `scripts` as well, but tell me if I am wrong.

**Review by jordanpadams — APPROVED** [review]

File: scratch/bad-doc-id-enumerator/enum.py (hunk 9 above)
Comment (jordanpadams) [reply scratch/bad-doc-id-enumerator/enum.py]: @alexdunnjpl ☝️ just wanted to loop back on this before we merge

Comment (alexdunnjpl) [reply scratch/bad-doc-id-enumerator/enum.py]: @jordanpadams sorry I overlooked this.

We can rename this (now or in future) if desired at your preference - the motive here was to communicate that the contents is trash (not to be relied upon and/or not part of the software), but _useful_ trash (worth keeping around until such time as they become important enough to treat like part of the software).

Granted, that's too disparaging for the aoss-reindexer and aoss-migration scripts, so if you want I can rename the dir prior to merge, and put a readme in the new scripts subdir communicating that the new bash scripts are rough dev utilities and not to be trusted or cared about?
